Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm

```diff
@@ -4944,12 +4944,28 @@ static VisiblePositionRange constrainRangeToSelection(const VisiblePositionRange
         }
     }
 
-    auto makeString = [] (const VisiblePosition& start, const VisiblePosition& end) -> AttributedString {
+    auto isTextObscured = [] (const VisiblePosition& visiblePosition) -> bool {
+        if (RefPtr textControl = enclosingTextFormControl(visiblePosition.deepEquivalent())) {
+            if (is<HTMLInputElement>(textControl)) {
+                auto& input = downcast<HTMLInputElement>(*textControl);
+                if (input.isAutoFilledAndObscured())
+                    return true;
+            }
+        }
+        return false;
+    };
+
+    auto makeString = [&isTextObscured] (const VisiblePosition& start, const VisiblePosition& end) -> AttributedString {
         auto range = makeSimpleRange(start, end);
         if (!range || range->collapsed())
             return { };
         // FIXME: This should return editing-offset-compatible attributed strings if that option is requested.
-        return WebCore::AttributedString::fromNSAttributedString(adoptNS([[NSAttributedString alloc] initWithString:WebCore::plainTextReplacingNoBreakSpace(*range, TextIteratorBehavior::EmitsOriginalText)]));
+
+        auto isObscured = isTextObscured(start) || isTextObscured(end);
```

Comment:
Nit - I think just checking only the start should be sufficient — `enclosingTextFormControl` should always return the same shadow host in both cases.

You could then inline the helper lambda above into this (though keeping them separate is okay too)